docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+    content: IyEvYmluL2Jhc2ggL...
+  - path: /etc/sysctl.d/99-k8s-general.conf
+    permissions: 0644
+    content: |
+      vm.max_map_count = 135217728
+      kernel.softlockup_panic = 1
+      kernel.softlockup_all_cpu_backtrace = 1
+      ...
+  - path: /opt/bin/health-monitor
+    permissions: 0755
+    content: |
+      #!/bin/bash
+      set -o nounset
+      set -o pipefail
+
+      function docker_monitoring {
+      ...
+status:
+  observedGeneration: ...
+  state: ...
+  lastError: ..
+  lastOperation: ...
+  cloudConfig: ...
+```
+
+Cloud-specific controllers which might need to add another kernel option or another flag to the kubelet, maybe even another file to the disk, can register a `MutatingWebhookConfiguration` to that resource and modify it upon creation/update.
+The task of the `MachineCloudConfig` controller is to only generate the OS-specific cloud-config based on the `.spec` field, but not to add or change any logic related to Shoots.
+
+##### Worker pools definition
+
+For every worker pool defined in the `Shoot` Gardener will create a `WorkerPool` CRD which shall be picked up by a cloud-specific controller and be translated to `MachineClass`es and `MachineDeployment`s.
+
+```yaml
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: WorkerPool
+metadata:
+  name: pool-01
+  namespace: shoot--core--aws-01
+spec:
+  cloudConfig: base64(downloader-cloud-config)
+  infrastructureProviderStatus:
+    apiVersion: aws.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureStatus
+    vpc:
+      id: vpc-1234
+      subnets:
+      - id: subnet-acbd1234
+        name: workers
+        zone: eu-west-1
+      securityGroups:
+      - id: sg-xyz12345
+        name: workers
+    iam:
+      nodesRoleARN: <some-arn>
+      instanceProfileName: foo
+    ec2:
+      keyName: bar
+  providerConfig:
+    apiVersion: aws.cloud.gardener.cloud/v1alpha1
+    kind: WorkerPoolConfig
+    ami: ami-d0dcef3b
+    machineType: m4.large
+    volumeType: gp2
+    volumeSize: 20Gi
+    zones:
+    - eu-west-1a
+  region: eu-west-1
+  secretRef:
+    name: my-aws-credentials
+  minimum: 2
+  maximum: 2
+status:
+  observedGeneration: ...
+  state: ...
+  lastError: ..
+  lastOperation: ...
+```
+
+#### Shoot state
+
+In order to be able to move the control plane of a Shoot between Seed clusters (e.g., if a Seed cluster is not available anymore or entirely broken) Gardener must store some non-reconstructable state, potentially also the state written by the controllers.
+Gardener watches over the extension CRDs and copies the `.status.state` in a `ShootState` resource in the Garden cluster.
+Any change to the status written by the respective CRD-controllers must be immediately reflected in the `ShootState` resource.
+The contract between Gardener and those controllers is: **Every controller must be capable of reconstructing its own environment based on both the state it has written before and on the real world's conditions/state.**
```

Comment:
> Why the shoot state information can't be stored in shoot CRs status field, and a new resource type is preferred?

Mainly for RBAC reasons (we might want to restrict who can interact with `ShootState` as it might contain secrets), and for not blowing up the `Shoot` resource.

> If we'd go for ab explicit CR, ShootStateBackup would be a better name since this resource is redundant and is here to provide data for restore functionality and not as a source of truth. I can imagine developers trying to consume this resource assuming this reflects the ShootState which would be wrong.

Nobody will be able to consume it, it's more an internal resource of Gardener and will not be readable/changeable from the outside.

Reply:
This worths mentioning this in the proposal since this seems to be the major reason for having a dedicated resource type.